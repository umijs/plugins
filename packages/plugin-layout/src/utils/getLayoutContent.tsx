import { LayoutConfig } from '../types/interface.d';

export default (
  userConfig: LayoutConfig,
  path: string,
  formatMessage: boolean,
) => `import React, { useState, useEffect } from "react";
import { ApplyPluginsType, useModel ${
  // 没有 formatMessage 就不打开国际化
  formatMessage ? `, useIntl` : ''
} } from "umi";
import { plugin } from "../core/umiExports";
import LayoutComponent from '${path}';

export default props => {
  const [runtimeConfig, setRuntimeConfig] = useState(null);

  const initialInfo = (useModel && useModel("@@initialState")) || {
    initialState: undefined,
    loading: false,
    setInitialState: null
  }; // plugin-initial-state 未开启

  useEffect(() => {
    const useRuntimeConfig =
      plugin.applyPlugins({
        key: "layout",
        type: ApplyPluginsType.modify,
        initialValue: initialInfo
      }) || {};
    if (useRuntimeConfig instanceof Promise) {
      useRuntimeConfig.then(config => {
        setRuntimeConfig(config);
      });
      return;
    }
    setRuntimeConfig(useRuntimeConfig);
  }, [initialInfo?.initialState]);

  const userConfig = {
    ...${JSON.stringify(userConfig).replace(/"/g, "'")},
    ...runtimeConfig || {}
  };

  ${formatMessage ? 'const { formatMessage } = useIntl();' : ''}

  if(!runtimeConfig){
    return null
  }

  return React.createElement(LayoutComponent, {
    userConfig,
    ${formatMessage ? 'formatMessage,' : ''}
    ...props
  });
};
`;
