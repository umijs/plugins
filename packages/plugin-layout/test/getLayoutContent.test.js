import getLayoutContent from '../src/utils/getLayoutContent';

describe('getLayoutContent', () => {
  test('getLayoutContent', () => {
    const LayoutContent = getLayoutContent({}, 'test');

    expect(
      LayoutContent.includes(`import LayoutComponent from 'test';`),
    ).toBeTruthy();
    expect(LayoutContent.includes(`...{}`)).toBeTruthy();
  });
});
