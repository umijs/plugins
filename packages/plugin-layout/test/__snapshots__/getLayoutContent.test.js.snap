// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLayoutContent genRenderRightContent 1`] = `
"import React from 'react';
  import { Avatar, Dropdown, Menu, Spin } from 'antd';
  import { SelectLang } from 'umi';
  import { LogoutOutlined } from '@ant-design/icons';
  import { ILayoutRuntimeConfig } from '../types/interface.d';

  export default function renderRightContent(
    runtimeLayout: ILayoutRuntimeConfig,
    loading: boolean,
    initialState: any,
    setInitialState: any,
  ) {
    if (runtimeLayout.rightRender) {
      return runtimeLayout.rightRender(
        initialState,
        setInitialState,
        runtimeLayout,
      );
    }

    const menu = (
      <Menu className=\\"umi-plugin-layout-menu\\">
        <Menu.Item
          key=\\"logout\\"
          onClick={() =>
            runtimeLayout.logout && runtimeLayout?.logout(initialState)
          }
        >
          <LogoutOutlined />
          退出登录
        </Menu.Item>
      </Menu>
    );

    const avatar = (
      <span className=\\"umi-plugin-layout-action\\">
        <Avatar
          size=\\"small\\"
          className=\\"umi-plugin-layout-avatar\\"
          src={
            initialState?.avatar ||
            'https://gw.alipayobjects.com/zos/antfincdn/XAosXuNZyF/BiazfanxmamNRoxxVxka.png'
          }
          alt=\\"avatar\\"
        />
        <span className=\\"umi-plugin-layout-name\\">{initialState?.name}</span>
      </span>
    );

    if (loading) {
      return (
        <div className=\\"umi-plugin-layout-right\\">
          <Spin size=\\"small\\" style={{ marginLeft: 8, marginRight: 8 }} />
        </div>
      );
    }

    return (
      <div className=\\"umi-plugin-layout-right anticon\\">
        {runtimeLayout.logout ? (
          <Dropdown overlay={menu} overlayClassName=\\"umi-plugin-layout-container\\">
            {avatar}
          </Dropdown>
        ) : (
          avatar
        )}
        {SelectLang && <SelectLang />}
      </div>
    );
  }
  "
`;

exports[`getLayoutContent genRenderRightContent initialState=false 1`] = `
"export default function renderRightContent() {
    return null;
  }
  "
`;

exports[`getLayoutContent genRenderRightContent locale=false 1`] = `
"import React from 'react';
  import { Avatar, Dropdown, Menu, Spin } from 'antd';
  
  import { LogoutOutlined } from '@ant-design/icons';
  import { ILayoutRuntimeConfig } from '../types/interface.d';

  export default function renderRightContent(
    runtimeLayout: ILayoutRuntimeConfig,
    loading: boolean,
    initialState: any,
    setInitialState: any,
  ) {
    if (runtimeLayout.rightRender) {
      return runtimeLayout.rightRender(
        initialState,
        setInitialState,
        runtimeLayout,
      );
    }

    const menu = (
      <Menu className=\\"umi-plugin-layout-menu\\">
        <Menu.Item
          key=\\"logout\\"
          onClick={() =>
            runtimeLayout.logout && runtimeLayout?.logout(initialState)
          }
        >
          <LogoutOutlined />
          退出登录
        </Menu.Item>
      </Menu>
    );

    const avatar = (
      <span className=\\"umi-plugin-layout-action\\">
        <Avatar
          size=\\"small\\"
          className=\\"umi-plugin-layout-avatar\\"
          src={
            initialState?.avatar ||
            'https://gw.alipayobjects.com/zos/antfincdn/XAosXuNZyF/BiazfanxmamNRoxxVxka.png'
          }
          alt=\\"avatar\\"
        />
        <span className=\\"umi-plugin-layout-name\\">{initialState?.name}</span>
      </span>
    );

    if (loading) {
      return (
        <div className=\\"umi-plugin-layout-right\\">
          <Spin size=\\"small\\" style={{ marginLeft: 8, marginRight: 8 }} />
        </div>
      );
    }

    return (
      <div className=\\"umi-plugin-layout-right anticon\\">
        {runtimeLayout.logout ? (
          <Dropdown overlay={menu} overlayClassName=\\"umi-plugin-layout-container\\">
            {avatar}
          </Dropdown>
        ) : (
          avatar
        )}
        
      </div>
    );
  }
  "
`;
